### ISSUE TITLE
Not possible to restrict egress traffic to the Kube API (and DNS) only

### ISSUE THREAD

**Issue by Cajga** [aws/eks-anywhere#4658]
**What happened**:
We want to use "default-deny" network policy so we installed the cluster with `policyEnforcementMode: "always"` config option in Cilium. This will enable policy enforcement by default on both Ingress and Egress traffic so without a network policy which allows a specific traffic, everything is blocked.

We have some kubernetes components which needs to access the Kubernetes API to subscribe to it's CRD resources. As such, we need to create egress network policies which allows these components to reach the API. One of these components is Sealed Secrets so I will use that as example here.

We tried to create kubernetes native network policy for sealed secrets so it can reach the Kubernetes API server (and the DNS) but it seems this is not possible (without opening all egress traffic). We have tried many type of policies but none of them worked.

1) First, we tried to filter based on IP. After multiple iteration we went as far as using 0.0.0.0/0 but this is still not working:
Used the following policy:
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-egress-to-dns-and-api
  namespace: sealed-secrets
spec:
  policyTypes:
    - Egress
  podSelector: {}
  egress:
    - to:
        - namespaceSelector: {}
          podSelector:
            matchLabels:
              k8s-app: kube-dns
      ports:
        - port: 53
          protocol: UDP
        - port: 53
          protocol: TCP
    - to:
        - ipBlock:
            cidr: 0.0.0.0/0
```

The policy does end up in the cilium policy (checked inside the cilium pod on the host which runs sealed secret pod):
```
# cilium policy get|grep -A 3 -B 3 0.0.0.0
      {
        "toCIDRSet": [
          {
            "cidr": "0.0.0.0/0"
          }
        ]
      }
```

However the traffic still gets dropped by cilium:
```
# cilium endpoint list|grep 3117 -A 6
3117       Enabled            Enabled           31447      k8s:app.kubernetes.io/instance=sealed-secrets                                                 100.80.3.116   ready   
                                                           k8s:app.kubernetes.io/name=sealed-secrets                                                                            
                                                           k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=sealed-secrets                                        
                                                           k8s:io.cilium.k8s.policy.cluster=default                                                                             
                                                           k8s:io.cilium.k8s.policy.serviceaccount=sealed-secrets                                                               
                                                           k8s:io.kubernetes.pod.namespace=sealed-secrets                                                                       
# cilium monitor --type drop 
Press Ctrl-C to quit
level=info msg="Initializing dissection cache..." subsys=monitor
xx drop (Policy denied) flow 0x37b1f956 to endpoint 0, , identity 31447->kube-apiserver: 100.80.3.116:60556 -> REMOVEDTHEIP:6443 tcp SYN
```

After further digging we found in the [documentation](https://docs.cilium.io/en/stable/policy/language/#ip-cidr-based) that it seems Cilium ignores IP based filters for traffic between endpoints which are both handled by Cilium itself:
>  CIDR rules do not apply to traffic where both sides of the connection are either managed by Cilium or use an IP belonging to a node in the cluster (including host networking pods).

2) As such, we tried to use label filters. After multiple iteration, we ended up trying to allow everything going to the kube-system namespace:
To check the labels on the namespace:
```
$ kubectl describe namespaces kube-system |grep -i labels
Labels:       kubernetes.io/metadata.name=kube-system
```

Then we used the following policy:
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-egress-to-dns-and-api
  namespace: sealed-secrets
spec:
  policyTypes:
    - Egress
  podSelector: {}
  egress:
    - to:
        - namespaceSelector: {}
          podSelector:
            matchLabels:
              k8s-app: kube-dns
      ports:
        - port: 53
          protocol: UDP
        - port: 53
          protocol: TCP
    - to:
        - namespaceSelector:
            matchLabels:
              kubernetes.io/metadata.name: kube-system      
          podSelector: {}
```

This also ended up in the cilium policy but it did not help (same deny message in the cilium monitor as before).

3) so just to verify if this was even possible we have opened up all egress traffic to the pods in the sealed secrets namespace:
```
$ cat allow-egress.yaml 
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-all-ingress-egress
  namespace: sealed-secrets
spec:
  podSelector: {}
  egress:
  - {}
  policyTypes:
  - Egress
```

This worked, selased secrets were able to start up and reach the API!

So we started digging a bit and it seems there is a relevant issue in Cilium github project which shows that it can be done with CiliumNetworkPolicy (ciliums own network policy CRD which is not supported on EKS-A) but no one managed with Kubernetes native network policy yet: https://github.com/cilium/cilium/issues/20550

**What you expected to happen**:
We can lock down the network following security best practices and only allow traffic to/from certain destinations

**How to reproduce it (as minimally and precisely as possible)**:
* see above

**Anything else we need to know?**:

**Environment**:
- EKS Anywhere Release: v0.13.1
- EKS Distro Release: 1.24

**Comment by sathieu** [comment]
Same as https://github.com/cilium/cilium/issues/20550?

**Comment by chrisdoherty4** [comment]
We've worked with Cilium developers to make this possible for EKS-A Cilium v1.13 which will be available in EKS-A v0.19 expected in Jan 2024.

**Comment by Cajga** [comment]
Hi @chrisdoherty4 thanks for the update. Great news!

**Comment by chrisdoherty4** [comment]
Resolved by #5030
